File: assets/importeer.js
/*
 * importeer.js — browserversie van importeer.py
 * ----------------------------------------------------------------------------
 * Zet een Obsidian-vriendelijk schrijfbestand (editie.md) om naar het
 * runtime-formaat (bron.txt), volledig client-side. Dezelfde logica als
 * importeer.py: schone hoofdtekst + een === NOTEN ===-blok met  lemma | code |
 * inhoud , dat per lemma in de tekst wordt verankerd; plus een optioneel
 * === REGISTER ===-blok dat als register:-regels aan de frontmatter komt.
 *
 * Gebruik:  var res = Importeer.naarBron(editieMdTekst);
 *           res.bron      -> de bron.txt-tekst
 *           res.rapport   -> [{status:'ok'|'mis', tekst, code}]
 */
(function (global) {
  'use strict';

  var NOTEN = /^\s*={3,}\s*NOTEN\s*={3,}\s*$/i;
  var REGISTER = /^\s*={3,}\s*REGISTER\s*={3,}\s*$/i;
  var LIJST_ITEM = /^([-*]|\d+\.)(\s+)(.*)$/;

  // ---- Frontmatter / hoofdtekst / noten / register ------------------------
  function parseBestand(tekst) {
    var m = tekst.match(/^---\s*\n([\s\S]*?)\n---\s*\n?/);
    var frontmatter = m ? m[1] : '';
    var rest = m ? tekst.slice(m[0].length) : tekst;

    var tekstregels = [], notenregels = [], registerregels = [];
    var sectie = 'tekst';
    rest.split('\n').forEach(function (regel) {
      if (NOTEN.test(regel)) { sectie = 'noten'; return; }
      if (REGISTER.test(regel)) { sectie = 'register'; return; }
      if (sectie === 'noten') notenregels.push(regel);
      else if (sectie === 'register') registerregels.push(regel);
      else tekstregels.push(regel);
    });

    var noten = [];
    notenregels.forEach(function (regel) {
      if (!regel.trim() || regel.trimStart().charAt(0) === '#' || regel.indexOf('|') === -1) return;
      var velden = splitPipe(regel, 3).map(function (x) { return x.trim(); });
      if (velden.length === 3) noten.push(velden);
    });

    var register = [];
    registerregels.forEach(function (regel) {
      if (!regel.trim() || regel.trimStart().charAt(0) === '#' || regel.indexOf('|') === -1) return;
      register.push(regel.trim());
    });
    return { frontmatter: frontmatter, hoofdtekst: tekstregels.join('\n'), noten: noten, register: register };
  }

  // Split op de eerste (max-1) pipes — als str.split met limiet, maar de rest
  // van de string blijft in het laatste veld (zoals Python split('|', 2)).
  function splitPipe(s, max) {
    var out = [], i = 0, start = 0;
    while (out.length < max - 1) {
      var p = s.indexOf('|', start);
      if (p === -1) break;
      out.push(s.slice(i, p));
      i = p + 1; start = p + 1;
    }
    out.push(s.slice(i));
    return out;
  }

  // Split op top-niveau | (respecteert geneste [[ ]]).
  function splitTopPipe(s) {
    var velden = [], cur = '', depth = 0, i = 0;
    while (i < s.length) {
      if (s.substr(i, 2) === '[[') { depth++; cur += '[['; i += 2; }
      else if (s.substr(i, 2) === ']]') { depth--; cur += ']]'; i += 2; }
      else if (s.charAt(i) === '|' && depth === 0) { velden.push(cur); cur = ''; i++; }
      else { cur += s.charAt(i); i++; }
    }
    velden.push(cur);
    return velden;
  }

  function tabelCellen(rij) {
    var s = rij.trim().replace(/^\|/, '').replace(/\|\s*$/, '');
    return splitTopPipe(s).map(function (c) { return c.trim(); });
  }
  function isScheidingsrij(rij) {
    var cellen = tabelCellen(rij);
    return cellen.length > 0 && cellen.every(function (c) { return /^:?-{3,}:?$/.test(c); });
  }

  // ---- Blokken ------------------------------------------------------------
  function maakBlokken(tekst) {
    var blokken = [];
    var para = [], lijst = [], tabel = [];

    function flushPara() {
      if (para.length) {
        blokken.push({ type: 'para', tekst: para.map(function (x) { return x.trim(); }).join(' ') });
        para = [];
      }
    }
    function flushLijst() {
      if (lijst.length) {
        var items = lijst.map(function (s) {
          var m = s.match(LIJST_ITEM);
          return { marker: m[1] + m[2], inhoud: m[3] };
        });
        blokken.push({ type: 'lijst', items: items, geordend: /^\d/.test(lijst[0]) });
        lijst = [];
      }
    }
    function flushTabel() {
      if (tabel.length) {
        var rijen = tabel.map(function (s) {
          var sep = isScheidingsrij(s);
          return { sep: sep, raw: s, cellen: sep ? [] : tabelCellen(s) };
        });
        blokken.push({ type: 'tabel', rijen: rijen });
        tabel = [];
      }
    }
    function flush() { flushPara(); flushLijst(); flushTabel(); }

    tekst.split('\n').forEach(function (regel) {
      var s = regel.trim();
      if (s === '') flush();
      else if (/^#{1,3}\s/.test(s)) { flush(); blokken.push({ type: 'kop', tekst: s }); }
      else if (/^!\[.*\]\(.*\)/.test(s)) { flush(); blokken.push({ type: 'fig', tekst: s }); }
      else if (s.charAt(0) === '~') { flush(); blokken.push({ type: 'mark', tekst: s }); }
      else if (/^@\s/.test(s)) { flush(); blokken.push({ type: 'dag', tekst: s }); }
      else if (LIJST_ITEM.test(s)) { flushPara(); flushTabel(); lijst.push(s); }
      else if (s.charAt(0) === '|') { flushPara(); flushLijst(); tabel.push(s); }
      else { flushLijst(); flushTabel(); para.push(s); }
    });
    flush();
    return blokken;
  }

  // ---- Lemma opzoeken -----------------------------------------------------
  function zoekSpan(tekst, delen, start, occ) {
    var lo = tekst.toLowerCase();
    if (delen.length === 1) {
      var deel = delen[0].toLowerCase();
      if (!deel) return null;
      if (occ) {
        var pos = 0, idx = -1;
        for (var n = 0; n < occ; n++) {
          idx = lo.indexOf(deel, pos);
          if (idx === -1) return null;
          pos = idx + deel.length;
        }
        return [idx, idx + deel.length];
      }
      var i = lo.indexOf(deel, start);
      return i === -1 ? null : [i, i + deel.length];
    }
    var kop = delen[0].toLowerCase(), staart = delen[delen.length - 1].toLowerCase();
    var hi = lo.indexOf(kop, start);
    if (hi === -1) return null;
    var ti = lo.indexOf(staart, hi + kop.length);
    if (ti === -1) return null;
    return [hi, ti + staart.length];
  }

  // ---- Segmenten (alinea / lijstitem / tabelcel) --------------------------
  function maakSegmenten(blokken) {
    var segs = [];
    blokken.forEach(function (b) {
      if (b.type === 'para') {
        segs.push({ lees: function () { return b.tekst; }, schrijf: function (v) { b.tekst = v; }, inserts: [] });
      } else if (b.type === 'lijst') {
        b.items.forEach(function (it) {
          segs.push({ lees: function () { return it.inhoud; }, schrijf: function (v) { it.inhoud = v; }, inserts: [] });
        });
      } else if (b.type === 'tabel') {
        b.rijen.forEach(function (rij) {
          if (rij.sep) return;
          rij.cellen.forEach(function (_c, ci) {
            segs.push({ lees: function () { return rij.cellen[ci]; }, schrijf: function (v) { rij.cellen[ci] = v; }, inserts: [] });
          });
        });
      }
    });
    return segs;
  }

  // ---- Verankeren ---------------------------------------------------------
  function ankerNoten(blokken, noten) {
    var segmenten = maakSegmenten(blokken);
    var rapport = [];
    var curSeg = 0, curPos = 0;

    noten.forEach(function (drietal) {
      var lemma = drietal[0], code = drietal[1], inhoud = drietal[2];
      var mm = lemma.match(/\((\d+)\)\s*$/);
      var occ = mm ? parseInt(mm[1], 10) : null;
      if (mm) lemma = lemma.slice(0, mm.index).trim();
      var delen = lemma.split(/\s*(?:…|\.\.\.)\s*/).map(function (d) { return d.trim(); });

      var gevonden = null;
      var si = curSeg, pos = curPos;
      while (si < segmenten.length) {
        var res = zoekSpan(segmenten[si].lees(), delen, pos, occ);
        if (res) { gevonden = [si, res]; break; }
        si++; pos = 0;
      }
      if (!gevonden) {
        for (var s2 = 0; s2 < segmenten.length; s2++) {
          var r2 = zoekSpan(segmenten[s2].lees(), delen, 0, occ);
          if (r2) { gevonden = [s2, r2]; break; }
        }
      }
      if (!gevonden) { rapport.push({ status: 'mis', tekst: lemma, code: code }); return; }
      var idx = gevonden[0], span = gevonden[1];
      var seg = segmenten[idx];
      seg.inserts.push({ s: span[0], e: span[1], code: code, inhoud: inhoud });
      curSeg = idx; curPos = span[1];
      rapport.push({ status: 'ok', tekst: seg.lees().slice(span[0], span[1]), code: code });
    });
    return { segmenten: segmenten, rapport: rapport };
  }

  // ---- Geneste [[ ]] bouwen -----------------------------------------------
  function bouwGenest(tekst, items) {
    items = items.slice().sort(function (a, b) { return a.s - b.s || b.e - a.e; });
    var roots = [], stack = [];
    items.forEach(function (it) {
      while (stack.length && it.s >= stack[stack.length - 1].e) stack.pop();
      if (stack.length) {
        var ouder = stack[stack.length - 1];
        if (it.e > ouder.e) return; // kruist de ouder: niet nestbaar
        (ouder.kids || (ouder.kids = [])).push(it);
      } else {
        roots.push(it);
      }
      stack.push(it);
    });
    function render(s, e, kids) {
      var uit = [], pos = s;
      (kids || []).slice().sort(function (a, b) { return a.s - b.s; }).forEach(function (k) {
        uit.push(tekst.slice(pos, k.s));
        var binnen = render(k.s, k.e, k.kids);
        uit.push('[[' + binnen + '|' + k.code + '|' + k.inhoud + ']]');
        pos = k.e;
      });
      uit.push(tekst.slice(pos, e));
      return uit.join('');
    }
    return render(0, tekst.length, roots);
  }

  function pasToe(segmenten) {
    segmenten.forEach(function (seg) {
      if (seg.inserts.length) seg.schrijf(bouwGenest(seg.lees(), seg.inserts));
    });
  }

  // ---- Bron opbouwen ------------------------------------------------------
  function bouwBron(frontmatter, blokken) {
    var uit = ['---', frontmatter, '---', ''];
    var vorig = null;
    blokken.forEach(function (b) {
      if (b.type === 'kop') {
        if (vorig !== null) uit.push('');
        uit.push(b.tekst); uit.push('');
      } else if (b.type === 'mark') {
        if (vorig === 'para') uit.push('');
        uit.push(b.tekst);
      } else if (b.type === 'dag') {
        if (vorig === 'para') uit.push('');
        uit.push(b.tekst);
      } else if (b.type === 'fig') {
        uit.push(''); uit.push(b.tekst); uit.push('');
      } else if (b.type === 'lijst') {
        uit.push('');
        b.items.forEach(function (it) { uit.push(it.marker + it.inhoud); });
        uit.push('');
      } else if (b.type === 'tabel') {
        uit.push('');
        b.rijen.forEach(function (rij) {
          uit.push(rij.sep ? rij.raw : '| ' + rij.cellen.join(' | ') + ' |');
        });
        uit.push('');
      } else {
        uit.push(b.tekst);
      }
      vorig = b.type;
    });
    return uit.join('\n').replace(/\s+$/, '') + '\n';
  }

  // ---- Hoofdfunctie -------------------------------------------------------
  function naarBron(editieMd) {
    var tekst = editieMd.replace(/%%[\s\S]*?%%/g, ''); // Obsidian-commentaar weg
    var p = parseBestand(tekst);
    var frontmatter = p.frontmatter;
    if (p.register.length) {
      var extra = p.register.map(function (r) { return 'register: ' + r; }).join('\n');
      frontmatter = frontmatter.replace(/\n+$/, '') + '\n' + extra;
    }
    var blokken = maakBlokken(p.hoofdtekst);
    var res = ankerNoten(blokken, p.noten);
    pasToe(res.segmenten);
    var bron = bouwBron(frontmatter, blokken);
    return { bron: bron, rapport: res.rapport };
  }

  global.Importeer = { naarBron: naarBron };
})(window);

File: importeer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
importeer.py — zet een Obsidian-vriendelijk schrijfbestand (editie.md) om naar
het runtime-formaat van de site (bron.txt).

Schrijfmodel (editie.md):
  - Frontmatter tussen ---, identiek aan de kop van bron.txt (titel, apparaat: ...).
  - Schone hoofdtekst: gewone alinea's (zachte regelafbrekingen worden samengevoegd),
    koppen (#/##/###), paginamarkeringen (~ label | doel) en typografie ({sc:}, {ab:} ...).
    GEEN nootmarkeringen in de tekst.
  - Een notenblok, ingeleid door een regel  === NOTEN ===  , met per regel:
        lemma | code | inhoud
    * lemma  = het woord of de woordgroep waar de noot bij hoort. Voor een lange
               passage: eerste woord … laatste woord  (met … of ...).
               Komt het lemma meer dan eens voor, kies dan met  lemma (2)  het
               tweede voorkomen, enzovoort.
    * code   = apparaatcode uit de kop (w, comm, krit, auteur, ...).
    * inhoud = de noottekst.

De importer zoekt elk lemma op in de hoofdtekst (in leesvolgorde, vanaf de vorige
noot) en schrijft er onze inline-noot [[lemma|code|inhoud]] van, in bron.txt.

Gebruik:  python3 importeer.py edities/van-der-meersch/editie.md
          (schrijft bron.txt naast het invoerbestand)
"""
import re
import sys
import os

NOTEN_SCHEIDING = re.compile(r'^\s*={3,}\s*NOTEN\s*={3,}\s*$', re.IGNORECASE)
REGISTER_SCHEIDING = re.compile(r'^\s*={3,}\s*REGISTER\s*={3,}\s*$', re.IGNORECASE)


def parse_bestand(tekst):
    """Splits frontmatter, hoofdtekst, notenblok en registerblok.

    Retour: (frontmatter, hoofdtekst, noten, registerregels). Het registerblok
    (=== REGISTER ===) bevat regels  soort | canonieke naam | variant, variant …
    die als  register:-regels aan de frontmatter worden toegevoegd."""
    m = re.match(r'^---\s*\n(.*?)\n---\s*\n?', tekst, re.DOTALL)
    frontmatter = m.group(1) if m else ''
    rest = tekst[m.end():] if m else tekst

    tekstregels, notenregels, registerregels = [], [], []
    sectie = 'tekst'
    for regel in rest.split('\n'):
        if NOTEN_SCHEIDING.match(regel):
            sectie = 'noten'; continue
        if REGISTER_SCHEIDING.match(regel):
            sectie = 'register'; continue
        if sectie == 'noten':
            notenregels.append(regel)
        elif sectie == 'register':
            registerregels.append(regel)
        else:
            tekstregels.append(regel)

    noten = []
    for regel in notenregels:
        if not regel.strip() or regel.lstrip().startswith('#') or '|' not in regel:
            continue
        velden = [x.strip() for x in regel.split('|', 2)]
        if len(velden) == 3:
            noten.append(tuple(velden))  # (lemma, code, inhoud)

    register = []
    for regel in registerregels:
        if not regel.strip() or regel.lstrip().startswith('#') or '|' not in regel:
            continue
        register.append(regel.strip())
    return frontmatter, '\n'.join(tekstregels), noten, register


LIJST_ITEM = re.compile(r'^([-*]|\d+\.)(\s+)(.*)$')


def split_top_pipe(s):
    """Split op |-tekens die NIET binnen geneste [[ ]] staan (zie editie.js)."""
    velden, cur, depth, i = [], '', 0, 0
    while i < len(s):
        if s[i:i + 2] == '[[':
            depth += 1; cur += '[['; i += 2
        elif s[i:i + 2] == ']]':
            depth -= 1; cur += ']]'; i += 2
        elif s[i] == '|' and depth == 0:
            velden.append(cur); cur = ''; i += 1
        else:
            cur += s[i]; i += 1
    velden.append(cur)
    return velden


def tabel_cellen(rij):
    s = rij.strip()
    s = re.sub(r'^\|', '', s)
    s = re.sub(r'\|\s*$', '', s)
    return [c.strip() for c in split_top_pipe(s)]


def is_scheidingsrij(rij):
    cellen = tabel_cellen(rij)
    return bool(cellen) and all(re.match(r'^:?-{3,}:?$', c) for c in cellen)


def maak_blokken(tekst):
    """Groepeer de hoofdtekst in blokken: alinea's (samengevoegd), koppen,
    markeringen, lijsten en tabellen."""
    blokken = []
    para, lijst, tabel = [], [], []

    def flush_para():
        if para:
            blokken.append({'type': 'para', 'tekst': ' '.join(x.strip() for x in para)})
            para.clear()

    def flush_lijst():
        if lijst:
            items = []
            for s in lijst:
                m = LIJST_ITEM.match(s)
                items.append({'marker': m.group(1) + m.group(2), 'inhoud': m.group(3)})
            blokken.append({'type': 'lijst', 'items': items,
                            'geordend': bool(re.match(r'^\d', lijst[0]))})
            lijst.clear()

    def flush_tabel():
        if tabel:
            rijen = []
            for s in tabel:
                sep = is_scheidingsrij(s)
                rijen.append({'sep': sep, 'raw': s,
                              'cellen': [] if sep else tabel_cellen(s)})
            blokken.append({'type': 'tabel', 'rijen': rijen})
            tabel.clear()

    def flush():
        flush_para(); flush_lijst(); flush_tabel()

    for regel in tekst.split('\n'):
        s = regel.strip()
        if s == '':
            flush()
        elif re.match(r'^#{1,3}\s', s):
            flush(); blokken.append({'type': 'kop', 'tekst': s})
        elif re.match(r'^!\[.*\]\(.*\)', s):
            flush(); blokken.append({'type': 'fig', 'tekst': s})
        elif s.startswith('~'):
            flush(); blokken.append({'type': 'mark', 'tekst': s})
        elif re.match(r'^@\s', s):
            flush(); blokken.append({'type': 'dag', 'tekst': s})
        elif LIJST_ITEM.match(s):
            flush_para(); flush_tabel(); lijst.append(s)
        elif s.startswith('|'):
            flush_para(); flush_lijst(); tabel.append(s)
        else:
            flush_lijst(); flush_tabel(); para.append(s)
    flush()
    return blokken


def zoek_span(tekst, delen, start, occ):
    """Vind de positie van een lemma in `tekst` vanaf `start` (hoofdletter-
    ongevoelig; de oorspronkelijke schrijfwijze in de tekst blijft behouden).
    `delen` = [heel] of [eerste, laatste] (bij een … in het lemma).
    `occ`   = gewenst voorkomen (1-based) of None."""
    lo = tekst.lower()
    if len(delen) == 1:
        deel = delen[0].lower()
        if occ:
            pos, idx = 0, -1
            for _ in range(occ):
                idx = lo.find(deel, pos)
                if idx == -1:
                    return None
                pos = idx + len(deel)
            return (idx, idx + len(deel))
        idx = lo.find(deel, start)
        return (idx, idx + len(deel)) if idx != -1 else None
    # eerste … laatste
    kop, staart = delen[0].lower(), delen[-1].lower()
    hi = lo.find(kop, start)
    if hi == -1:
        return None
    ti = lo.find(staart, hi + len(kop))
    if ti == -1:
        return None
    return (hi, ti + len(staart))


class Segment:
    """Een doorzoekbaar, muteerbaar stuk tekst: een alinea, een lijstitem of
    een tabelcel. `lees`/`schrijf` koppelen het aan zijn plek in het blok."""
    def __init__(self, lees, schrijf):
        self._lees, self._schrijf = lees, schrijf
        self.inserts = []

    @property
    def tekst(self):
        return self._lees()

    def commit(self):
        if self.inserts:
            self._schrijf(bouw_genest(self.tekst, self.inserts))


def maak_segmenten(blokken):
    """Vlakke lijst van doorzoekbare segmenten, in leesvolgorde."""
    segs = []
    for b in blokken:
        if b['type'] == 'para':
            segs.append(Segment(lambda b=b: b['tekst'],
                                lambda v, b=b: b.__setitem__('tekst', v)))
        elif b['type'] == 'lijst':
            for it in b['items']:
                segs.append(Segment(lambda it=it: it['inhoud'],
                                    lambda v, it=it: it.__setitem__('inhoud', v)))
        elif b['type'] == 'tabel':
            for rij in b['rijen']:
                if rij['sep']:
                    continue
                for ci in range(len(rij['cellen'])):
                    segs.append(Segment(
                        (lambda rij=rij, ci=ci: rij['cellen'][ci]),
                        (lambda v, rij=rij, ci=ci: rij['cellen'].__setitem__(ci, v))))
    return segs


def anker_noten(blokken, noten):
    """Verankert elke noot in het juiste segment (alinea, lijstitem of tabelcel),
    in leesvolgorde vanaf de vorige noot, met terugval naar het begin."""
    segmenten = maak_segmenten(blokken)
    rapport = []
    cur_seg, cur_pos = 0, 0

    for lemma, code, inhoud in noten:
        m = re.search(r'\((\d+)\)\s*$', lemma)
        occ = int(m.group(1)) if m else None
        if m:
            lemma = lemma[:m.start()].strip()
        delen = [d.strip() for d in re.split(r'\s*(?:…|\.\.\.)\s*', lemma)]

        gevonden = None
        si, pos = cur_seg, cur_pos
        while si < len(segmenten):
            res = zoek_span(segmenten[si].tekst, delen, pos, occ)
            if res:
                gevonden = (si, res)
                break
            si, pos = si + 1, 0
        if not gevonden:  # terugval: vanaf het begin
            for si2 in range(len(segmenten)):
                res = zoek_span(segmenten[si2].tekst, delen, 0, occ)
                if res:
                    gevonden = (si2, res)
                    break

        if not gevonden:
            rapport.append(('NIET GEVONDEN', lemma, code))
            continue
        si, (s, e) = gevonden
        seg = segmenten[si]
        seg.inserts.append(dict(s=s, e=e, code=code, inhoud=inhoud))
        cur_seg, cur_pos = si, e
        rapport.append(('ok', seg.tekst[s:e], code))
    return segmenten, rapport


def pas_toe(segmenten):
    """Wikkel de gevonden spans in [[lemma|code|inhoud]] (geneste noten worden
    correct in elkaar genest; echt kruisende noten worden gemeld)."""
    for seg in segmenten:
        seg.commit()


def bouw_genest(tekst, items):
    # Sorteer: buitenste eerst (kleinste start, grootste eind).
    items = sorted(items, key=lambda x: (x['s'], -x['e']))
    roots, stack = [], []
    for it in items:
        while stack and it['s'] >= stack[-1]['e']:
            stack.pop()
        if stack:
            ouder = stack[-1]
            if it['e'] > ouder['e']:  # kruist de ouder: niet nestbaar
                sys.stderr.write(f"  ! kruisende (overlappende) noot overgeslagen: "
                                 f"{tekst[it['s']:it['e']]!r}\n")
                continue
            ouder.setdefault('kids', []).append(it)
        else:
            roots.append(it)
        stack.append(it)

    def render(s, e, kids):
        uit, pos = [], s
        for k in sorted(kids, key=lambda x: x['s']):
            uit.append(tekst[pos:k['s']])
            binnen = render(k['s'], k['e'], k.get('kids', []))
            uit.append(f"[[{binnen}|{k['code']}|{k['inhoud']}]]")
            pos = k['e']
        uit.append(tekst[pos:e])
        return ''.join(uit)

    return render(0, len(tekst), roots)


def bouw_bron(frontmatter, blokken):
    uit = ['---', frontmatter, '---', '']
    vorig = None
    for b in blokken:
        if b['type'] == 'kop':
            if vorig is not None:
                uit.append('')
            uit.append(b['tekst'])
            uit.append('')
        elif b['type'] == 'mark':
            if vorig == 'para':
                uit.append('')
            uit.append(b['tekst'])
        elif b['type'] == 'dag':
            if vorig == 'para':
                uit.append('')
            uit.append(b['tekst'])
        elif b['type'] == 'fig':
            uit.append('')
            uit.append(b['tekst'])
            uit.append('')
        elif b['type'] == 'lijst':
            uit.append('')
            for it in b['items']:
                uit.append(it['marker'] + it['inhoud'])
            uit.append('')
        elif b['type'] == 'tabel':
            uit.append('')
            for rij in b['rijen']:
                uit.append(rij['raw'] if rij['sep'] else '| ' + ' | '.join(rij['cellen']) + ' |')
            uit.append('')
        else:  # para
            uit.append(b['tekst'])
        vorig = b['type']
    return '\n'.join(uit).rstrip() + '\n'


def main():
    if len(sys.argv) < 2:
        sys.exit('gebruik: python3 importeer.py <editie.md>')
    invoer = sys.argv[1]
    with open(invoer, encoding='utf-8') as f:
        tekst = f.read()

    # Obsidian-commentaar %% ... %% wordt genegeerd (bv. voor een legenda).
    tekst = re.sub(r'%%[\s\S]*?%%', '', tekst)

    frontmatter, hoofdtekst, noten, register = parse_bestand(tekst)
    if register:
        extra = '\n'.join('register: ' + r for r in register)
        frontmatter = frontmatter.rstrip('\n') + '\n' + extra
    blokken = maak_blokken(hoofdtekst)
    segmenten, rapport = anker_noten(blokken, noten)
    pas_toe(segmenten)
    bron = bouw_bron(frontmatter, blokken)

    uitvoer = os.path.join(os.path.dirname(invoer), 'bron.txt')
    with open(uitvoer, 'w', encoding='utf-8') as f:
        f.write(bron)

    ok = sum(1 for r in rapport if r[0] == 'ok')
    mis = [r for r in rapport if r[0] != 'ok']
    print(f'{ok}/{len(rapport)} noten verankerd  ->  {uitvoer}')
    for r in mis:
        print(f'  NIET GEVONDEN: {r[1]!r} ({r[2]})')


if __name__ == '__main__':
    main()
